contracts/Router.sol

```diff
@@ -98,21 +104,14 @@ contract Router is IRouter, EIP712, FeeGenerator {
     /// @param user The user address
     /// @return The calculated agent address for the user
     function calcAgent(address user) external view returns (address) {
-        address result = address(
-            uint160(
-                uint256(
-                    keccak256(
-                        abi.encodePacked(
-                            bytes1(0xff),
-                            address(this),
-                            bytes32(bytes20(user)),
-                            keccak256(abi.encodePacked(type(Agent).creationCode, abi.encode(agentImplementation)))
-                        )
-                    )
-                )
-            )
+        // https://github.com/matter-labs/v2-testnet-contracts/blob/main/l2/contracts/L2ContractHelper.sol
+        bytes32 constructorInputHash = keccak256(abi.encode(agentImplementation));
```

Comment:
which one is close to the `immutable` here?
1.  can add as an `immutable` variable in global and assign value in the `constructor`
2.  use `calldata` to make sure this variable is read-only

 
 If 1, I am still figuring out why this needs to be in the `constructor`
 If 2, I believe `calldata` can not be used for basic type like `bytes32`
 Or maybe I missed a 3 here?

Reply:
IMO 1 can reduce the calculation since that `agentImplementation` is decided in constructor.

Reply:
I seeee. But do we need to save the calculation if this is only for an external view function?

Reply:
I think it depends on how the function is used. If someone calls this function in a function that writes state, then the gas consumption matters. Keeping this might saves some gas on deployment. I prefer doing this through assigning a immutable value, but completely fine with the current design. It's up to you.

Reply:
fixed